crates/runtime/src/instance/allocator/pooling/memory_pool.rs

```diff
@@ -87,78 +170,128 @@ impl MemoryPool {
-        // the same size as guard regions for other memories.
-        let allocation_size = memory_and_guard_size
-            .checked_mul(max_total_memories)
-            .and_then(|c| c.checked_add(initial_memory_offset))
-            .ok_or_else(|| {
-                anyhow!("total size of memory reservation exceeds addressable memory")
-            })?;
-
-        // Create a completely inaccessible region to start
-        let mapping = Mmap::accessible_reserved(0, allocation_size)
+        let layout = calculate(&constraints)?;
+        log::debug!("creating memory pool: {constraints:?} -> {layout:?}");
+        let mut mapping = Mmap::accessible_reserved(0, layout.total_slab_bytes)
             .context("failed to create memory pool mapping")?;
 
+        // Then, stripe the memory with the available protection keys. This is
+        // unnecessary if there is only one stripe color.
+        if layout.num_stripes >= 2 {
+            let mut cursor = layout.pre_slab_guard_bytes;
+            let pkeys = &pkeys[..layout.num_stripes];
+            for i in 0..constraints.num_memory_slots {
+                let pkey = &pkeys[i % pkeys.len()];
+                let region = unsafe { mapping.slice_mut(cursor..cursor + layout.slot_bytes) };
+                pkey.protect(region)?;
+                cursor += layout.slot_bytes;
+            }
+            debug_assert_eq!(
+                cursor + layout.post_slab_guard_bytes,
+                layout.total_slab_bytes
+            );
+        }
+
         let image_slots: Vec<_> = std::iter::repeat_with(|| Mutex::new(None))
-            .take(max_total_memories)
+            .take(constraints.num_memory_slots)
             .collect();
 
-        let pool = Self {
-            index_allocator: ModuleAffinityIndexAllocator::new(
-                config.limits.total_memories,
+        let create_stripe = |i| {
+            let num_slots = constraints.num_memory_slots / layout.num_stripes
+                + usize::from(constraints.num_memory_slots % layout.num_stripes > i);
+            let allocator = ModuleAffinityIndexAllocator::new(
+                num_slots.try_into().unwrap(),
                 config.max_unused_warm_slots,
-            ),
+            );
+            Stripe {
+                allocator,
+                pkey: pkeys.get(i).cloned(),
+            }
+        };
+
+        debug_assert!(layout.num_stripes > 0);
+        let stripes: Vec<_> = (0..layout.num_stripes)
+            .into_iter()
+            .map(create_stripe)
+            .collect();
+
+        let pool = Self {
+            stripes,
             mapping,
             image_slots,
-            memory_size: memory_size.try_into().unwrap(),
-            memory_and_guard_size,
-            initial_memory_offset,
-            max_total_memories,
+            memory_size: constraints.max_memory_bytes,
+            memory_and_guard_size: layout.slot_bytes,
+            pre_slab_guard_size: layout.pre_slab_guard_bytes,
```

Comment:
Should `Self` perhaps store the whole `layout` and maybe `constraints` while we're at it? (reducing duplication and keeping things close to their source of truth)

Reply:
I was tempted to do that but then worried that it would take up a few extra bytes. If you're ok with that, I'll move those in and maybe remove some of the fields that they duplicate.